Extract repeated scripts into files

marcomorain opened the pull request:
The current pytorch config.yml is causing some backend performance
problems on CircleCI, due to the size of the file when all of the YAML
anchors have been expanded. You can view the "processed" config as our
internal system deal with it by running `circleci config process`.

    circleci config process .circleci/config.yml | wc -c

    Before: 2833769 bytes
    After:   558252 bytes (~80% less)

Add create a new job, `setup`, that has 2 functions:
- Assert that config.yml is up to date
- Put the .circleci/scripts directory into a workspace, so that
downstream jobs can easily access it.

The `setup` job becomes the parent of all jobs in the workflow. This
allows us to fail fast if config is invalid. It might be a good place to
add other, quick, lint checks to help fail the build faster.

yf225 commented:
@kostmo Would you like to review this PR?

marcomorain commented:
There is an unforseen problem with this approach – several of the jobs don't have a `checkout` step that runs before `setup_ci_environment`, so the extracted script file is not present on disk when the step runs.

I guess one option would be to add a `checkout` to these steps, but that's not ideal.

kostmo reviewed (commented):
@pjh5 just landed some substantive refactoring in #18888; you'd better take a look at the new `master` in case there's overlap.

kostmo commented:
> There is an unforseen problem with this approach – several of the jobs don't have a `checkout` step that runs before `setup_ci_environment`, so the extracted script file is not present on disk when the step runs.

@ezyang, @pjh5, and I discussed a method of `curl` piped to `sh` ([demo](https://github.com/kostmo/curl-and-execute-demo/blob/master/bootstrap.sh#L5)) to execute whichever offloaded scripts need to be run before a checkout is present.  I don't recall how far that idea was taken in #18888.

pjh5 commented:
@marcomorain we should be able to do something like https://github.com/pytorch/pytorch/blob/master/.circleci/verbatim-sources/nightly-binary-build-defaults.yml#L27-L41 , to duplicate the git checkout logic (which commit to check out) and then curl the script without needing to checkout everything first.

marcomorain commented on .circleci/scripts/setup_linux_system_environment.sh:
```diff
@@ -0,0 +1,40 @@
+# TODO: https://circleci.com/gh/pytorch/pytorch/1478441
+# The first `grep` here is exiting with exit code 1 when running in CI.
+# The output of `ps auxfww` does not match `[a]pt`.
+# ps auxfww | grep '[a]pt'
```
@kostmo I'm having a very strange issue with this `grep`. When I extracted the script out to a file, this started to fail. `grep` exits with code 1 (no output matched).

Any ideas what's up?

glenjamin replied:
I think these are in there to try and detect an issue we were having with a stray `apt` process leftover from the VM provisionining.

You shouldn't be getting failures though, I can't see a `-o pipefail` anywhere.

marcomorain replied:
Pipefail might be inherited differently from the parent shell (the shell the build step is executed in) when invoking the script in this way.

marcomorain commented:
This is ready for a proper review now, please @kostmo 🙏 

cc @glenjamin

pjh5 commented on .circleci/verbatim-sources/job-specs-setup.yml:
```diff
@@ -0,0 +1,12 @@
+    docker:
+      - image: circleci/python:3.7.3
+    steps:
+      - checkout
```
This replaces the location of the scripts for all the binaries from this checkout instead of https://github.com/pytorch/pytorch/blob/master/.circleci/scripts/binary_checkout.sh#L21-L41 . Will they always be the same?

pjh5 replied:
For context, I added our own checkout for all our binary jobs cause they check out two repos, not just one, but I wasn't trying to do anything magical.

glenjamin replied:
The checkout change wasn't intended to be anything special - the main thing we did was use the workspace to share the scripts around.

The setup job persists the scripts that any job will need into the workspace, and then all the other jobs can pick them up from there instead of having to `checkout` or `curl` with the retry.

pjh5 replied:
I don't think the files will always be the same. For example this PR https://github.com/pytorch/pytorch/pull/19765 would start a bunch of jobs that checkout an arbitrary pytorch branch. By changing binary_checkout.sh we can point to that branch directly. But if all the jobs use the built-in checkout then they would all use the same commit the CircleCI is running on.

glenjamin replied:
I'm not quite following what the goal is there, if you wanted to run against an arbitrary branch, can you open a pull request for that branch to have it start building?

CircleCI usually builds every branch unless there's a branch filter in the workflow config, but this project is configured to only build when a PR exists, see "Advanced Settings":
>By default, we will build all the commits for this project. Once turned on, we will only build branches that have associated pull requests open. 

<hr />

We haven't intentionally changed to not use binary_checkout in any of the "real" jobs, we only added `checkout` to the new `setup` job, which is used to provide the `.circleci/scripts` directory to the other jobs.

The scripts directory that gets passed around in `~/workspace` would be the one from the PR branch, but the rest of the jobs remain free to do whatever they would like.

glenjamin replied:
We haven't currently changed any of the existing steps which use curl to download a script file - we've only used the workspace mechanism for files which were previously inlined. Based on my understanding of how things fit together, I believe this approach should be a suitable replacement for every scenario where you currently curl a file down from github if you wanted to move over.

pjh5 replied:
I am following this up in another comment below.

pjh5 commented on .circleci/verbatim-sources/linux-binary-build-defaults.yml:
```diff
@@ -27,13 +29,11 @@ binary_linux_build: &binary_linux_build
       no_output_timeout: "1h"
-      command: |
-        source "/pytorch/.circleci/scripts/binary_linux_build.sh"
+      command: ~/workspace/.circleci/scripts/binary_linux_build.sh
```
@glenjamin my comment should really be placed here. If we keep this change then this other checkout of Pytorch https://github.com/pytorch/pytorch/blob/master/.circleci/scripts/binary_checkout.sh#L20-L38 should be removed, since the this job will call into the scripts downloaded in the circleci-checkout instead of these.

Now that I look at it more, would this work for the nightly case https://github.com/pytorch/pytorch/blob/master/.circleci/scripts/binary_checkout.sh#L34-L36 when we want to build the last Pytorch branch of the previous day rather than the current master?

glenjamin replied:
Aha, I'm with you now. You're right - this script is now building from the wrong checkout.

As this job has already run `binary_checkout` by this point, we don't need to use the script from the workspace.

i'll go back over the PR and check the script paths more closely.

glenjamin commented:
@pjh5 I've rebased this to resolve the conflicts and correct the script you spotted that was using the wrong source.

I've gone back over the rest of the diff and I believe we're now correctly only using the workspace for scripts which don't require a specific working tree.

Could you give this another review and let us know if there are any other blockers to landing?

Thanks!

pjh5 commented on .circleci/verbatim-sources/nightly-binary-build-defaults.yml:
```diff
@@ -26,70 +26,23 @@
-    script="$workdir/pytorch/.circleci/scripts/binary_populate_env.sh"
-    cat "$script"
-    source "$script"
+  command: ~/workspace/.circleci/scripts/binary_populate_env.sh
```
In all the binary jobs, binary_checkout runs first, followed by other binary_ prefixed steps. With your changes, after binary_checkout runs there should be two pytorch checkouts, one at ~/workspace and one at $workdir/pytorch, where $workdir is / on dockers, /home/cirleci/project on machine executors, and /Users/distiller/project on mac executors. I think all the binary_ prefixed steps need to use the $workdir/pytorch checkout instead of the one in ~/workspace.

The binary jobs run at three schedules; there's a few that run on every PR, a few that run on commits merged to master, and a few that run at 9pm PST every day. We need all the binary jobs that run at 9pm every day to use the exact same commit, even if they don't start until a few minutes later. It is not rare for people on the west coast to be landing things at this time, so we can't just pull the latest master in every binary job since it can race. We have this line https://github.com/pytorch/pytorch/blob/master/.circleci/scripts/binary_checkout.sh#L35 in the binary checkout to force the exact same commit across all the binary jobs. This is why we need to use the $workdir/pytorch checkout in the binary jobs.

I think an alternative approach would be to have some parent job that clones master and persists it to the workspace, and then have all the binary jobs require that parent job. But that will take some refactoring.

glenjamin replied:
I'll have to digest this a bit more tomorrow (I'm on UK time), but one thing that springs to mind is that we could record the git SHA in the `setup` job, and have every downstream job pick that up.

I'll read over the `binary_` prefxied scripts more carefully - it's worth noting that the checkout in `~/workspace` will only contain the `.circleci/scripts` dir, which should make it fast to share.

pjh5 replied:
I've thought about this a little more. I think our options are

1. call binary_checkout from ~/workspace and the rest of the binary_* steps from $workdir. This won't take a lot of work but it is unexpected and'll be confusing to people in the future.
2. Create a binary_setup step, that saves the git SHA of master when it runs, then have all the binary_* jobs require binary_setup. I think this will still require a custom special step in all the binary_* jobs, as it'll have to respect the saved git SHA.
3. Leave the binary jobs as is for now (don't use the new setup step at all in them). Take the huge saving from eliminating setup_ci_env , and look at slimming down the binary_* steps further in a future PR. The binary_* steps are rather slim right now anyways. If we still see problems then I can look into doing #2 later.

glenjamin replied:
I've had a look through the `binary_` scripts, they all appear to me to not be location-sensitive, so if they were run from the ~/workspace they would function correctly - is the concern that any changes to these scripts might be inconsistent?

> We need all the binary jobs that run at 9pm every day to use the exact same commit, even if they don't start until a few minutes later. 

I double-checked this with our workflows team, all jobs created as part of a workflow will run with the same CIRCLE_SHA1, so you shouldn't need to do anything extra to acheive this.

I picked a couple of jobs out of the most [recent nightly workflow](https://circleci.com/workflow-run/5d9cc1ab-117b-4e68-a4da-33757160624a)
* https://circleci.com/gh/pytorch/pytorch/1552842
* https://circleci.com/gh/pytorch/pytorch/1553291

You can see from the "Spin up environment" step that these both have `CIRCLE_SHA1` populated to the same ref, and from the output in "Checkout" we can see that the `elif [[ -n "$CIRCLE_SHA1" ]]; then` conditional branch is being used.

Based on my understanding of the scripts I believe that should make the current proposed changes safe - but I could well still be missing something.

-----------

I'm happy to look at implementing option 3 if you'd prefer as that should make the review more straightforward.

pjh5 replied:
If that's the case then can you remove the extra Pytorch checkout logic in binary_checkout so that only one pytorch is checked out? Then the rest of these changes should be fine.

glenjamin committed:
Extract repeated scripts into files

The current pytorch config.yml is causing some backend performance
problems on CircleCI, due to the size of the file when all of the YAML
anchors have been expanded. You can view the "processed" config as our
internal system deal with it by running `circleci config process`.

    circleci config process .circleci/config.yml | wc -c

Before: 2181979 bytes
After:   452044 bytes (~80% less)

1. Add create a new job, `setup`, that has 2 functions:
- Assert that `config.yml` is up to date
- Put the `.circleci/scripts` directory into a workspace, so that
downstream jobs can easily access it.

The `setup` job becomes the parent of all jobs in the workflow. This
allows us to fail fast if config is invalid. It might be a good place to
add other, quick, lint checks to help fail the build faster.

2. Attach a workspace at ~/workspace in any jobs that need to access
`.circleci/scripts`.

3. Remove a part of the binary_checkout script that tried to sync up which
commit was used on scheduled builds because CircleCI already does this for us

glenjamin commented:
@pjh5 I've rebased the branch again, I believe everything should be in order now.

glenjamin commented:
Looks like I overlooked the way binary_upload uses the workspace to share the built binaries (it seems to use a slightly different method to binary_test, although I can't quite figure out what's different)

I've pushed up a PR which restores the other workspace attachment to keep things working how they did before: https://github.com/pytorch/pytorch/pull/20265
